src/os_posix/os_union_fs.c

```diff
@@ -536,168 +540,139 @@ __union_fs_file_lock(WT_FILE_HANDLE *file_handle, WT_SESSION *wt_session, bool l
-                    __wt_verbose_debug3(session, WT_VERB_FILEOPS, "EXTENT MATCH %s, no work done",
-                      union_fh->iface.name);
-                    return (0);
-                } else {
-                    /* Otherwise grow the extent to match the new write. */
-                    WT_ASSERT(session, alloc->len < len);
-                    alloc->len = len;
-                    if (alloc->next != NULL && alloc->next->off <= EXTENT_END(alloc)) {
-                        WT_RET(__union_merge_with_next_extents(session, alloc));
-                    }
-                    return (0);
-                }
-            }
-            if (alloc->off > offset) {
-                break;
-            }
-            prev = alloc;
-            alloc = alloc->next;
-        }
+    hole = union_fh->destination.hole_list;
+    while (hole != NULL) {
 
-        if (prev == NULL) {
-            // New extent is at the start.
-            /* Allocate a new extent. */
-            WT_RET(__wt_calloc_one(session, &new));
-            new->off = offset;
-            new->len = len;
-            new->next = alloc;
-            union_fh->destination.allocation_list = new;
-            return (0);
-
-        } else if (alloc == NULL) {
-            // New extent is after.
-            if (EXTENT_END(prev) == offset) {
-                // Grow prev.
-                prev->len += len;
-                return (0);
-            }
-        } else {
-            // New extent is in between.
-            if (EXTENT_END(prev) == offset) {
-                // Grow prev.
-                prev->len += len;
-                if (prev->next->off <= EXTENT_END(prev)) {
-                    __union_merge_with_next_extents(session, prev);
-                }
-                return (0);
-            }
-        }
-        /* Allocate a new extent. */
-        WT_RET(__wt_calloc_one(session, &new));
-        new->off = offset;
-        new->len = len;
-        new->next = alloc;
-        prev->next = new;
-        if (EXTENT_END(prev) >= new->off) {
-            __union_merge_with_next_extents(session, prev);
+        if (read_end < hole->off)
+            /* All subsequent holes are past the read. We won't find matching holes */
+            break;
+
+        read_begins_in_hole = ADDR_IN_EXTENT(offset, hole);
+        read_ends_in_hole = ADDR_IN_EXTENT(read_end, hole);
+        if (read_begins_in_hole && read_ends_in_hole) {
+            /* Our read is entirely within a hole */
+            __wt_verbose_debug3(session, WT_VERB_FILEOPS,
+              "CANNOT SERVICE %s: Reading from hole. Read: %ld-%ld, hole: %ld-%ld",
+              union_fh->iface.name, offset, read_end, hole->off, EXTENT_END(hole));
+            return (false);
+        } else if (read_begins_in_hole != read_ends_in_hole) {
+            /*
+             * The read starts in a hole but doesn't finish in it, or vice versa. This should never
+             * happen.
```

Comment:
Instead of "This should never happen" I would say something like: "This breaks assumptions that this code makes about how the block manager works and is intentionally unimplemented."

Reply:
Done in [9f8e50e](https://github.com/wiredtiger/wiredtiger/pull/11268/commits/9f8e50e9e9e25c96dd296661234032fd94c5c45c)